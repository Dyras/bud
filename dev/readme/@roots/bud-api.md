## Overview

Surfaces various configuration utilities for use in end-user configs and extensions.

[See the documentation](https://bud.js.org/docs/api.alias)
