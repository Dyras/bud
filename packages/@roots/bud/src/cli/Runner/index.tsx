import {boundMethod as bind} from 'autobind-decorator'
import {isFunction} from 'lodash'

import {config, Framework} from '../../..'
import {Factory} from '../../Factory'
import {Config} from '../Config'

class Runner {
  public app: Framework

  public options: {
    mode?: 'production' | 'development'
    config?: Partial<config>
  }

  public cli: {
    flags: {[key: string]: any}
    args: any[]
  }

  public mode: 'development' | 'production'

  /**
   * Fluent builders
   * @example `bud.config.{js,ts}`
   */
  public _fluentBuilders: string[]
  public get fluentBuilders(): string[] {
    return this._fluentBuilders
  }
  public set fluentBuilders(builders: string[]) {
    this._fluentBuilders = builders
  }

  /**
   * Fluent builders by env
   * @example `bud.development.{js,ts}`
   */
  public _fluentBuildersByEnv: string[]
  public get fluentBuildersByEnv(): string[] {
    return this._fluentBuildersByEnv
  }
  public set fluentBuildersByEnv(builders: string[]) {
    this._fluentBuildersByEnv = builders
  }

  /**
   * Static builders
   * @example `bud.config.{json,yml}`
   */
  public _staticBuilders: string[]
  public get staticBuilders(): string[] {
    return this._staticBuilders
  }
  public set staticBuilders(builders: string[]) {
    this._staticBuilders = builders
  }

  /**
   * Static builders by env
   * @example `bud.development.{json,yml}`
   */
  public _staticBuildersByEnv: string[]
  public get staticBuildersByEnv(): string[] {
    return this._staticBuildersByEnv
  }
  public set staticBuildersByEnv(builders: string[]) {
    this._staticBuildersByEnv = builders
  }

  /**
   * Class constructor
   *
   * @param cli
   * @param options
   * @param app
   */
  public constructor(
    cli: Runner['cli'],
    options: Runner['options'] = {},
    app: Framework = null,
  ) {
    Object.assign(this, {cli, options})
    this.setEnv(
      options?.mode || this.cli.flags.mode || 'production',
    )

    this.app =
      app ??
      Factory({
        mode: this.mode,
        ...options,
        config: {
          ...config,
          ...(options.config ?? {}),
          cli: this.cli.flags.cli ? false : true,
        },
      })

    Object.assign(this, {
      fluentBuilders: [
        `${this.app.name}.ts`,
        `${this.app.name}.js`,
        `${this.app.name}.config.ts`,
        `${this.app.name}.config.js`,
      ],
      fluentBuildersByEnv: [
        `${this.app.name}.${this.app.mode}.ts`,
        `${this.app.name}.${this.app.mode}.js`,
        `${this.app.name}.${this.app.mode}.config.ts`,
        `${this.app.name}.${this.app.mode}.config.js`,
      ],
      staticBuilders: [
        `${this.app.name}.json`,
        `${this.app.name}.yaml`,
        `${this.app.name}.yml`,
        `${this.app.name}.config.json`,
        `${this.app.name}.config.yaml`,
        `${this.app.name}.config.yml`,
      ],
      staticBuildersByEnv: [
        `${this.app.name}.${this.app.mode}.json`,
        `${this.app.name}.${this.app.mode}.yaml`,
        `${this.app.name}.${this.app.mode}.yml`,
        `${this.app.name}.${this.app.mode}.config.json`,
        `${this.app.name}.${this.app.mode}.config.yaml`,
        `${this.app.name}.${this.app.mode}.config.yml`,
      ],
    })
  }

  @bind
  public async make(build = true) {
    this.cli.flags.install && this.app.project.peers.install()

    this.cli.flags.discover &&
      this.app.project.peers.registerDiscovered()

    await this.doStatics()
    await this.doBuilders()

    if (build) {
      this.cli.flags.cache && this.app.persist()
      this.cli.flags.cache &&
        this.app.children.every((_name, child) =>
          child.persist(),
        )

      if (this.cli.flags.minimize) {
        this.app.minimize()

        this.app.children.every((_name, child) => {
          child.minimize()
        })
      }

      /**
       * Target was specified
       * @example `bud build --target plugin`
       */
      if (this.cli.flags.target.length > 0) {
        /**
         * Handle parent if applicable
         */
        !this.cli.flags.target.includes('bud') &&
          this.app.hooks.on('build/entry', false)
        /**
         * And children if applicable
         */
        this.app.children.getKeys().forEach(name => {
          !this.cli.flags.target.includes(name) &&
            this.app.children.remove(name)
        })
      }
    }

    return this.app
  }

  /**
   * Actually calls config fn and
   * configures application instance
   * @param configs
   */
  @bind
  public async build(configs: string[]): Promise<void> {
    const builder = await new Config(this.app, configs).get()
    isFunction(builder) && builder(this.app)
  }

  @bind
  public setEnv(env: 'production' | 'development') {
    process.env.BABEL_ENV = env
    process.env.NODE_ENV = env
  }

  @bind
  public async doBuilders() {
    await this.build(this.fluentBuilders)
    await this.build(this.fluentBuildersByEnv)
  }

  @bind
  public async doStatics() {
    await new Config(this.app, this.staticBuilders).apply()

    await new Config(this.app, this.staticBuildersByEnv).apply()
  }
}

export {Runner}
