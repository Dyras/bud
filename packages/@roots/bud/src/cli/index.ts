/**
 * Mandatory export for @oclif/cli
 */
export {run} from '@oclif/command'
