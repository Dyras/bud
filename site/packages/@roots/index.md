---
id: 'packages'
title: '@roots/bud packages'
slug: '/'
sidebar_label: 'Packages'
sidebar_position: 0.5
custom_edit_url: null
---
