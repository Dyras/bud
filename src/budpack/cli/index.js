import React, {useEffect, useState} from 'react'
import notifier from 'node-notifier'

import useWebpack from './hooks/useWebpack'
import useFocusState from './hooks/useFocusState'

import BrowserSync from './components/BrowserSync'
import Errors from './components/Errors'
import Assets from './components/Assets'
import App from './components/App'

/**
 * Budpack build status display
 *
 * @prop {object} compiler webpack compiler
 * @prop {string} mode watch or run
 */
const BudpackCLI = ({compiler, mode}) => {
  const [state, actions] = useFocusState()
  const build = useWebpack({compiler, mode})
  const [errors, setErrors] = useState(null)
  useEffect(() => {
    build?.errors && setErrors(build.errors)
  }, [build])

  useEffect(() => {
    !errors?.length > 0 &&
      build?.percentage == 1 &&
      notifier.notify({
        title: 'Build complete',
        message: build?.assets?.length > 0 ? `${build?.assets?.length} assets built.` : `Assets built.`,
      })
  }, [build?.percentage, errors])

  return (
    <App build={build} errors={errors} state={state} mode={mode}>
      <Assets actions={actions} build={build} />
      <Errors actions={actions} errors={errors} />
      <BrowserSync actions={actions} />
    </App>
  )
}

export default BudpackCLI
