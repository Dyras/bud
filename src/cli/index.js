const React = require('react')
const {useEffect, useState} = React
const PropTypes = require('prop-types')
const {Text, Box, Color} = require('ink')
const webpack = require('webpack')
const build = require('./../build').webpack
const compiler = webpack(build())

/**
 * Budpack CLI interface
 *
 * @prop {object} cli
 */
const Budpack = ({cli}) => {
  const [mode, setMode] = useState(null)
  useEffect(() => {
    cli && cli.input && setMode(cli.input[0])
  }, [cli])

  const [stats, setStats] = useState(null)
  const [error, setError] = useState(null)
  useEffect(() => {
    const callback = (err, stats) => {
      ;(err || stats.hasErrors()) && setError(err)
      setStats(stats)
    }

    if (mode) {
      mode == 'build' && compiler.run(callback)
      mode == 'dev' && compiler.watch(null, callback)
    }
  }, [mode])

  const [assets, setAssets] = useState([])
  useEffect(() => {
    stats &&
      stats.compilation &&
      stats.compilation.assets &&
      setAssets(Object.keys(stats.compilation.assets))
  }, [stats])

  return (
    <Box flexDirection="column" marginTop={1}>
      <Box flexDirection="column">
        <Text>
          <Color green>Budpack</Color>
        </Text>
        {mode && <Text>Target: {mode}</Text>}
      </Box>

      {assets &&
        assets.map((asset, id) => (
          <Box key={id}>
            <Text>{asset}</Text>
          </Box>
        ))}

      {assets && mode == 'dev' && (
        <Box marginTop={2} marginBottom={2}>
          ðŸ‘€ Watching the days go by.
        </Box>
      )}

      {error && console.log(error)}
    </Box>
  )
}

Budpack.propTypes = {
  cli: PropTypes.object,
}

module.exports = Budpack
