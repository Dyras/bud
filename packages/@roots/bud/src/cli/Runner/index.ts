import {Configuration} from '@roots/bud-framework'
import {bind} from '@roots/bud-support'
import {isUndefined} from 'lodash'

import {Bud} from '../../Bud'
import {config} from '../../config'
import {factory} from '../../factory'
import * as dynamic from './dynamic.config'
import * as flags from './flags.config'
import * as manifest from './manifest.config'

/**
 * @internal
 */
export class Runner {
  /**
   * @public
   */
  public app: Bud

  /**
   * Class constructor
   *
   * @param cli - CLI state
   * @param options - Bud options
   */
  public constructor(public cli: Configuration['cli']) {}

  /**
   * Initialize bud application
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public async initialize() {
    const determineDefaultBool = (value, fallback = true) =>
      isUndefined(value) ? fallback : value === true

    const settings = {
      config: {
        cli: this.cli,
        mode: this.cli.flags.mode ?? 'production',
        location: {
          project:
            this.cli.flags['location.project'] ??
            config.location.project,
          src:
            this.cli.flags['location.src'] ??
            config.location.src,
          dist:
            this.cli.flags['location.dist'] ??
            config.location.dist,
          storage:
            this.cli.flags['location.storage'] ??
            config.location.storage,
        },
        cache: {
          type:
            this.cli.flags['cache.type'] ?? config.cache.type,
        },
        features: {
          cache: determineDefaultBool(this.cli.flags.cache),
          clean: determineDefaultBool(this.cli?.flags.clean),
          dashboard: determineDefaultBool(
            this.cli.flags.dashboard,
          ),
          hash: determineDefaultBool(this.cli.flags.hash, false),
          html: determineDefaultBool(this.cli.flags.html, false),
          inject: determineDefaultBool(this.cli.flags.inject),
          install: determineDefaultBool(
            this.cli.flags.install,
            false,
          ),
          log: determineDefaultBool(this.cli.flags.log, false),
          manifest: determineDefaultBool(
            this.cli.flags.manifest,
          ),
          splitChunks: determineDefaultBool(
            this.cli.flags.splitChunks,
            false,
          ),
        },
      },
    }

    this.app = await factory(settings)

    this.app.dump(settings, {prefix: 'cli settings'})

    this.app.logger.instance
      .scope('cli', 'runner')
      .success('framework pre-compilation lifecycle complete.')

    this.app.dump(
      this.app,
      {
        prefix: 'bud post lifecycle',
        printFunctionName: true,
        language: 'typescript',
        callToJSON: false,
      },
      3,
    )

    return this.app
  }

  /**
   * Main process
   *
   * @param build - Boolean value indicating if compilation should occur
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public async make() {
    try {
      await dynamic.configs(this.app)

      await manifest.configs(this.app, 'configs.json.global')
      await manifest.configs(
        this.app,
        'configs.json.conditional',
      )
    } catch (error) {
      this.app.error(error)
    }

    await flags.config(this.app, this.cli.flags)

    return this.app
  }
}
