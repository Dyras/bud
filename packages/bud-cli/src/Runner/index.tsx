import {React, render, Instance} from '@roots/bud-support'
import Compile from '../containers/Compile'
import {Serve} from '../containers/Serve'
import type {Framework} from '@roots/bud-typings'

/**
 * ## bud.cli
 *
 * Ink application controller.
 */
export class Runner {
  /** Bud reference */
  public bud: Framework

  /**
   * ## bud.cli.instance
   *
   * Ink instance
   */
  public instance: Instance

  /**
   * Class constructor.
   */
  constructor(bud: Framework) {
    this.bud = bud
    this.run = this.run.bind(this)
  }

  public init(): void {
    return
  }

  /**
   * Run the compilation.
   */
  public run = function (): void {
    this.instance = this.bud.mode.is('development')
      ? render(<Serve bud={this.bud} />)
      : render(<Compile bud={this.bud} />)
  }

  public kill = function (): void {
    this.instance.quit()
  }
}
