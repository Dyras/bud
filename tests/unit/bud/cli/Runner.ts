/* eslint-disable no-console */
import {Runner} from '@roots/bud/src/cli/Runner'
import {Framework} from '@roots/bud-framework'

import {mocks} from './__mocks__/index'

describe('Runner', () => {
  let runner: Runner
  let bud: Framework

  beforeEach(async () => {
    bud = mocks.bud as unknown as Framework
    runner = new Runner(
      {args: [], argv: [], flags: {}, raw: [], metadata: {}},
      {},
      bud,
    )
  })

  afterEach(() => {
    jest.restoreAllMocks()
    jest.clearAllMocks()
  })

  it('make', () => {
    expect(runner.make).toBeInstanceOf(Function)
  })

  it('build', () => {
    expect(runner.build).toBeInstanceOf(Function)
  })

  it('setEnv', () => {
    expect(runner.setEnv).toBeInstanceOf(Function)
  })

  it('app property matches mock', () => {
    expect(runner.app).toMatchSnapshot(bud)
  })

  it('setEnv sets process.env.NODE_ENV', () => {
    runner.setEnv('production')
    expect(process.env.NODE_ENV).toBe('production')
  })

  it('setEnv sets process.env.BABEL_ENV', () => {
    runner.setEnv('development')
    expect(process.env.BABEL_ENV).toBe('development')
  })
})
