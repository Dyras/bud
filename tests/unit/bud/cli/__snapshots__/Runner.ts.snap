// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runner app property matches mock 1`] = `
Object {
  "cache": Object {
    "updateProfile": [MockFunction],
  },
  "close": [MockFunction],
  "dashboard": Object {
    "render": [MockFunction],
  },
  "entry": [MockFunction],
  "mode": "production",
  "name": "bud",
  "project": Object {
    "getValues": [MockFunction],
    "has": [MockFunction],
    "hasPeerDependency": [MockFunction],
    "peers": Object {
      "install": [MockFunction],
    },
  },
  "tap": [MockFunction],
  "use": [MockFunction],
}
`;
